### PR TITLE
ChipTactic and KickTactic using FSMs

### LINKED ISSUE #1865: Create get behind ball tactic
### Description of the task
Currently KickAction and ChipAction repeat code to get behind the ball. This should be pulled into another tactic so the code can be shared

### Acceptance criteria

- [ ] Create get behind ball action
- [ ] Refactor uses in chiptactic

### LINKED ISSUE #1923: Make a composite FSM Tactic
### Description of the task
Make a composite FSM Tactic example and path find for future composite FSMs

### Acceptance criteria

- [ ] composite FSM Tactic
- [ ] guide future composite FSMs

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
First composite FSM with ChipTactic and KickTactic's FSMs using the GetBehindBallFSM.

### Testing Done
FSM unit tests, will add simulated tests in #1922

### Resolved Issues
resolves #1923
resolves #1865

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/kick_fsm_test.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    fsm.process_event(KickFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::state<KickFSM::kick_state>));
+    robot = ::TestUtil::createRobotAtPos(Point(-2, 1.8));
+    world =
+        ::TestUtil::setBallVelocity(world, Vector(0, -2.1), Timestamp::fromSeconds(123));
+    EXPECT_TRUE(world.ball().hasBallBeenKicked(Angle::threeQuarter()));
+    fsm.process_event(KickFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::X));
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/kick_fsm_test.cpp]: This is kind of hard to follow, would you be able to add some comments here + newlines

File: src/software/util/sml_fsm/dump_sml_fsm.hpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
+#pragma once
```
Comment (akhilveeraghanta) [inline src/software/util/sml_fsm/dump_sml_fsm.hpp]: would be nice to have some more docs in this file

**Review by akhilveeraghanta — COMMENTED** [review]
looking sweet! will do a deeper dive once other members have had a chance to review

File: src/software/util/sml_fsm/dump_sml_fsm.hpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/util/sml_fsm/dump_sml_fsm.hpp]: I didn't mean to include this, whoops.

File: src/software/ai/hl/stp/tactic/kick_fsm_test.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/kick_fsm_test.cpp]: added comments

**Review by itsarune — COMMENTED** [review]
looks good, wanted a little clarification

File: src/software/ai/hl/stp/tactic/get_behind_ball_tactic.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#include "software/ai/hl/stp/tactic/get_behind_ball_tactic.h"
+
+#include <algorithm>
+
+GetBehindBallTactic::GetBehindBallTactic(bool loop_forever)
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/get_behind_ball_tactic.cpp:5]: are get_behind_ball_tactic and get_behind_ball_fsm basically duplicates using different frameworks?

**Review by itsarune — COMMENTED** [review]
Ah disregard the other comment, I got confused by the file names. The tactic folder is looking a bit bloated, what about putting the isms into a separate fsm folder within tactic

**Comment by jonathanlew** [comment]
> Ah disregard the other comment, I got confused by the file names. The tactic folder is looking a bit bloated, what about putting the isms into a separate fsm folder within tactic

We're going to move each tactic and their fsms into their own folder soon

**Review by itsarune — APPROVED** [review]
👍👍

**Review by akhilveeraghanta — APPROVED** [review]
:fire: this is awesome
the FSMs are super readable! nice work
